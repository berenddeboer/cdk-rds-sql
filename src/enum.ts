export enum RdsSqlResource {
  SCHEMA = "schema",
  ROLE = "role",
  SQL = "sql",
  DATABASE = "database",
  PARAMETER_PASSWORD = "parameter_password",
}
