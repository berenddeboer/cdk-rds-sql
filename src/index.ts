export * from "./enum"
export * from "./provider"
export * from "./sql"
export * from "./schema"
export * from "./role"
export * from "./database"
export * from "./iam-grant"
