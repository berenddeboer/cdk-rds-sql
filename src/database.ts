import { CustomResource } from "aws-cdk-lib"
import { Construct } from "constructs"
import { RdsSqlResource } from "./enum"
import { IProvider, DatabaseEngine } from "./provider"
import { Role } from "./role"

interface DatabaseAttributes {
  /**
   * Name of database to create.
   */
  readonly databaseName: string

  /**
   * Optional database owner.
   */
  readonly owner?: Role
}

export interface DatabaseProps extends DatabaseAttributes {
  /**
   * Provider.
   */
  readonly provider: IProvider
}

export interface IDatabase {
  readonly databaseName: string
}

class ImportedDatabase extends Construct implements IDatabase {
  public readonly databaseName: string

  constructor(scope: Construct, id: string, props: DatabaseAttributes) {
    super(scope, id)
    this.databaseName = props.databaseName
  }
}

export class Database extends CustomResource implements IDatabase {
  /**
   * Return a Database based upon name only. Use for importing existing databases.
   */
  static fromDatabaseName(scope: Construct, id: string, databaseName: string): IDatabase {
    return new ImportedDatabase(scope, id, {
      databaseName: databaseName,
    })
  }

  public readonly databaseName: string

  constructor(scope: Construct, id: string, props: DatabaseProps) {
    // Check if using DSQL provider and forbid database creation
    if (props.provider.engine === DatabaseEngine.DSQL) {
      throw new Error(
        "Database creation is not supported with DSQL. DSQL always uses 'postgres' database."
      )
    }

    super(scope, id, {
      serviceToken: props.provider.serviceToken,
      properties: {
        Resource: RdsSqlResource.DATABASE,
        ResourceId: props.databaseName,
        ...(props.provider.secret ? { SecretArn: props.provider.secret.secretArn } : {}),
        Owner: props.owner?.roleName,
      },
    })
    this.node.addDependency(props.provider)
    this.databaseName = props.databaseName
    if (props.owner) {
      this.node.addDependency(props.owner)
    }
  }
}
