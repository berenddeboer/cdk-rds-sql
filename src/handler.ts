import {
  SecretsManagerClient,
  GetSecretValueCommand,
  GetSecretValueCommandOutput,
} from "@aws-sdk/client-secrets-manager"
import {
  CloudFormationCustomResourceCreateEvent,
  CloudFormationCustomResourceUpdateEvent,
  CloudFormationCustomResourceDeleteEvent,
} from "aws-lambda"
import { backOff } from "exponential-backoff"
import { format } from "node-pg-format"
import { Client } from "pg"
import { RdsSqlResource } from "./enum"

interface CustomResourceResponse {
  PhysicalResourceId?: string
  Data?: any
  NoEcho?: boolean
}

type CreateResourceHandler = (
  resourceId: string,
  props?: any
) => Promise<string | string[]>
type UpdateResourceHandler = (
  resourceId: string,
  oldResourceId: string,
  props?: any
) => Promise<string | string[] | void>
type DeleteResourceHandler = (resourceId: string, props?: any) => string | string[] | void

type JumpTable = {
  [key in RdsSqlResource]: {
    Create: CreateResourceHandler
    Update: UpdateResourceHandler
    Delete: DeleteResourceHandler
  }
}

interface RoleProps {
  /**
   * Secret ARN.
   */
  PasswordArn: string

  /**
   * Optional database to which role is granted connect permissions.
   */
  DatabaseName?: string
}

interface DatabaseProps {
  Owner: string
}

interface DatabaseUpdateProps extends DatabaseProps {
  MasterOwner: string
}

const maxAttempts = 20

const jumpTable: JumpTable = {
  sql: {
    Create: (_: string, props?: any) => {
      return props.Statement
    },
    Update: (_: string, __: string, props?: any) => {
      return props.Statement
    },
    Delete: (_: string, props?: any) => {
      return props.Rollback
    },
  },
  schema: {
    Create: async (resourceId: string) => {
      return format("create schema if not exists %I", resourceId)
    },
    Update: async (resourceId: string, oldResourceId: string) => {
      return format("alter schema %I rename to %I", oldResourceId, resourceId)
    },
    Delete: (resourceId: string) => {
      return format("drop schema if exists %I cascade", resourceId)
    },
  },
  role: {
    Create: async (resourceId: string, props: RoleProps) => {
      if (!props.PasswordArn) throw "No PasswordArn provided"
      const password = await getPassword(props.PasswordArn)
      if (password) {
        const sql = [
          "start transaction",
          format("create role %I with login password %L", resourceId, password),
        ]
        if (props.DatabaseName) {
          // grant connect to database if database already exists
          sql.push(
            format(
              `DO $$
BEGIN
  IF EXISTS (select from pg_database where datname = '%s' and datistemplate = false) THEN
    grant connect on database %I to %I;
  END IF;
END$$;`,
              props.DatabaseName,
              props.DatabaseName,
              resourceId
            )
          )
        }
        sql.push("commit")
        return sql
      } else {
        throw `Cannot parse password from ${props.PasswordArn}`
      }
    },
    Update: async (resourceId: string, oldResourceId: string, props: RoleProps) => {
      // TODO: if database name has changed in OldResourceProperties, revoke connect
      if (props && props.PasswordArn) {
        const password = await getPassword(props.PasswordArn)
        if (password) {
          const sql = ["start transaction"]
          if (oldResourceId !== resourceId) {
            sql.push(format("alter role %I rename to %I", oldResourceId, resourceId))
          }
          sql.push(format("alter role %I with password %L", resourceId, password))
          if (props.DatabaseName) {
            sql.push(
              format("grant connect on database %I to %I", props.DatabaseName, resourceId)
            )
          }
          sql.push("commit")
          return sql
        } else {
          throw `Cannot parse password from ${props.PasswordArn}`
        }
      } else {
        const sql = ["start transaction"]
        if (oldResourceId !== resourceId) {
          sql.push(format("alter role %I rename to %I", oldResourceId, resourceId))
        }
        sql.push(
          format(
            `DO $$
BEGIN
  IF EXISTS (select from pg_database where datname = '%s' and datistemplate = false) THEN
    grant connect on database %I to %I;
  END IF;
END$$;`,
            props.DatabaseName,
            props.DatabaseName,
            resourceId
          )
        )
        sql.push("commit")
        return sql
      }
    },
    Delete: (resourceId: string, props: RoleProps) => {
      // TODO: if user is owner of a database, assign ownership to master user
      // This will require a specified inheritor on role creation
      return [
        "start transaction",
        format(
          `DO $$
BEGIN
  IF EXISTS (select from pg_catalog.pg_roles WHERE rolname = '%s') AND EXISTS (select from pg_database WHERE datname = '%s') THEN
    revoke all privileges on database %I from %I;
  END IF;
END$$;`,
          resourceId,
          props.DatabaseName,
          props.DatabaseName,
          resourceId
        ),
        format("drop role if exists %I", resourceId),
        "commit",
      ]
    },
  },
  database: {
    Create: async (resourceId: string, props: DatabaseProps) => {
      const owner = props.Owner
      if (owner) {
        return [
          format("create database %I", resourceId),
          format("alter database %I owner to %I", resourceId, owner),
        ]
      } else {
        return format("create database %I", resourceId)
      }
    },
    Update: async (
      resourceId: string,
      oldResourceId: string,
      props: DatabaseUpdateProps
    ): Promise<string[]> => {
      const statements: string[] = []
      if (resourceId !== oldResourceId) {
        if (props.MasterOwner) {
          statements.push(
            format("alter database %I owner to %I", oldResourceId, props.MasterOwner)
          )
        }
        statements.push(
          format("alter database %I rename to %I", oldResourceId, resourceId)
        )
      }
      const owner = props.Owner
      if (owner) {
        statements.push(format("alter database %I owner to %I", resourceId, props.Owner))
      }
      return statements
    },
    Delete: (resourceId: string, newOwner: string) => {
      return [
        format(
          "select pg_terminate_backend(pg_stat_activity.pid) from pg_stat_activity where datname = %L",
          resourceId
        ),
        format(
          "DO $$BEGIN\nIF EXISTS (select from pg_database WHERE datname = '%s') THEN alter database %I owner to %I; END IF;\nEND$$;",
          resourceId,
          resourceId,
          newOwner
        ),
        format("drop database if exists %I", resourceId),
      ]
    },
  },
}

const log =
  process.env.LOGGER === "true"
    ? console.debug
    : (_message?: any, ..._optionalParams: any[]) => {}

export const handler = async (
  event:
    | CloudFormationCustomResourceCreateEvent
    | CloudFormationCustomResourceUpdateEvent
    | CloudFormationCustomResourceDeleteEvent
): Promise<any> => {
  log(event)

  const requestType = event.RequestType
  const resource: RdsSqlResource = event.ResourceProperties.Resource
  const resourceId = event.ResourceProperties.ResourceId
  const databaseName = event.ResourceProperties.DatabaseName

  if (!Object.keys(jumpTable).includes(event.ResourceProperties.Resource)) {
    throw `Resource type '${resource}' not recognised.`
  }

  const secrets_client = new SecretsManagerClient({})
  const command = new GetSecretValueCommand({
    SecretId: event.ResourceProperties.SecretArn,
  })
  // As the IAM credentials can be cached, an update makde very recent
  // could not yet be available.
  // So we retry this a bit.
  log("Fetching secret")
  const secret: GetSecretValueCommandOutput = await backOff(
    async () => {
      try {
        const result = await secrets_client.send(command)
        return result
      } catch (e) {
        log("Error fetching secret %o", e)
        throw e
      }
    },
    {
      numOfAttempts: 10,
      startingDelay: 500,
    }
  )
  if (!secret.SecretString) throw "No secret string"
  const secretValues = JSON.parse(secret.SecretString)

  let sql: string | string[] | void
  switch (requestType) {
    case "Create": {
      sql = await jumpTable[resource][requestType](resourceId, event.ResourceProperties)
      break
    }
    case "Update": {
      const oldResourceId = (event as CloudFormationCustomResourceUpdateEvent)
        .PhysicalResourceId
      sql = await jumpTable[resource][requestType](resourceId, oldResourceId, {
        ...event.ResourceProperties,
        MasterOwner: secretValues.username,
      })
      break
    }
    case "Delete": {
      if (resource === RdsSqlResource.DATABASE) {
        sql = jumpTable[resource][requestType](resourceId, secretValues.username)
      } else {
        sql = jumpTable[resource][requestType](resourceId, event.ResourceProperties)
      }
      break
    }
  }

  if (sql) {
    let database: string
    if (resource === RdsSqlResource.ROLE) {
      database = secretValues.dbname
    } else {
      database = databaseName ?? secretValues.dbname // connect to given database if possible, else to database mentioned in secret
    }
    const params = {
      host: secretValues.host,
      port: secretValues.port,
      user: secretValues.username,
      password: secretValues.password,
      database: database,
      connectionTimeoutMillis: 2000, // return an error if a connection could not be established within 2 seconds
    }
    log(
      `Connecting to host ${params.host}: ${params.port}, database ${params.database} as ${params.user}`
    )
    log("Executing SQL", sql)
    const pg_client = new Client(params)
    await pg_client.connect()
    try {
      await backOff(
        async () => {
          if (typeof sql === "string") {
            return pg_client.query(sql)
          } else {
            if (sql) {
              return Promise.all(
                sql.map((statement) => {
                  return pg_client.query(statement)
                })
              )
            } else {
              return
            }
          }
        },
        {
          retry: errorFilter,
          numOfAttempts: maxAttempts,
        }
      )
    } finally {
      await pg_client.end()
    }
  }

  let response: CustomResourceResponse = {}
  // Except for the SQL resource, return the new resource id. This
  // will cause a delete to be send for the old resource.
  if (resource !== RdsSqlResource.SQL) {
    response.PhysicalResourceId = resourceId
  }

  return response
}

// Custom error filter, mainly to retry role creation.
// Frequently see "tuple concurrently updated", and adding
// dependencies is very hard to make work.
const errorFilter = (error: any, nextAttemptNumber: number) => {
  // Retry only if the error message contains "tuple concurrently"
  // This will cover concurrent updates and deletes
  const willRetry = error.message.includes("tuple concurrently")
  log(
    "Encountered an error on attempt %d/%d retry=%s error=[%o]",
    nextAttemptNumber - 1,
    maxAttempts,
    willRetry,
    error
  )
  return willRetry
}

/**
 * Parse password field from secret. Returns void on error.
 */
const getPassword = async (arn: string): Promise<string | void> => {
  const secrets_client = new SecretsManagerClient({})
  const command = new GetSecretValueCommand({
    SecretId: arn,
  })
  const secret = await secrets_client.send(command)
  if (secret.SecretString) {
    const json = JSON.parse(secret.SecretString)
    return json.password
  }
}
