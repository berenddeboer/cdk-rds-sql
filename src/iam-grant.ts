import { CustomResource } from "aws-cdk-lib"
import { Construct } from "constructs"
import { RdsSqlResource } from "./enum"
import { IProvider, DatabaseEngine } from "./provider"
import * as crypto from "crypto"

export interface IamGrantProps {
  /**
   * Provider.
   */
  readonly provider: IProvider

  /**
   * Database role name to grant IAM access to.
   */
  readonly roleName: string

  /**
   * IAM resource ARN (role, user, or other IAM principal).
   */
  readonly resourceArn: string
}

export class IamGrant extends CustomResource {
  private static createUniqueResourceId(roleName: string, resourceArn: string): string {
    const hash = crypto
      .createHash("sha256")
      .update(resourceArn)
      .digest("hex")
      .substring(0, 8)
    return `${roleName}:${hash}`
  }

  constructor(scope: Construct, id: string, props: IamGrantProps) {
    // IAM grants are only supported on DSQL
    if (props.provider.engine !== DatabaseEngine.DSQL) {
      throw new Error(
        "IAM grants are only supported with DSQL clusters. Use regular database authentication for RDS/Aurora clusters."
      )
    }

    super(scope, id, {
      serviceToken: props.provider.serviceToken,
      resourceType: "Custom::DsqlIamGrant",
      properties: {
        Resource: RdsSqlResource.IAM_GRANT,
        ResourceId: IamGrant.createUniqueResourceId(props.roleName, props.resourceArn),
        RoleName: props.roleName,
        ResourceArn: props.resourceArn,
      },
    })
    this.node.addDependency(props.provider)
  }
}
