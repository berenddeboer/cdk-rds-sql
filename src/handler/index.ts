import {
  CloudFormationCustomResourceCreateEvent,
  CloudFormationCustomResourceUpdateEvent,
  CloudFormationCustomResourceDeleteEvent,
} from "aws-lambda"
import { handler as handlerImpl } from "./handler.js"

export async function handler(
  event:
    | CloudFormationCustomResourceCreateEvent
    | CloudFormationCustomResourceUpdateEvent
    | CloudFormationCustomResourceDeleteEvent
): Promise<any> {
  return handlerImpl(event)
}
