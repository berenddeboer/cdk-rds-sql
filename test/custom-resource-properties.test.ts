import * as cdk from "aws-cdk-lib"
import { Match, Template } from "aws-cdk-lib/assertions"
import * as dsql from "aws-cdk-lib/aws-dsql"
import * as ec2 from "aws-cdk-lib/aws-ec2"
import * as rds from "aws-cdk-lib/aws-rds"
import { Provider, DatabaseEngine, Role, Database, Schema, Sql } from "../src"

describe("Custom Resource Properties", () => {
  let app: cdk.App
  let stack: cdk.Stack
  let vpc: ec2.Vpc
  let cluster: rds.DatabaseCluster
  let provider: Provider

  beforeEach(() => {
    app = new cdk.App()
    stack = new cdk.Stack(app, "TestStack", {
      env: {
        account: "123456789012",
        region: "us-east-1",
      },
    })

    vpc = new ec2.Vpc(stack, "Vpc", {
      subnetConfiguration: [
        {
          cidrMask: 28,
          name: "rds",
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    })

    cluster = new rds.DatabaseCluster(stack, "Cluster", {
      engine: rds.DatabaseClusterEngine.auroraPostgres({
        version: rds.AuroraPostgresEngineVersion.VER_15_10,
      }),
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      defaultDatabaseName: "example",
      writer: rds.ClusterInstance.serverlessV2("writer", {
        instanceIdentifier: "writer",
        publiclyAccessible: false,
        enablePerformanceInsights: false,
      }),
      serverlessV2MinCapacity: 0.5,
      serverlessV2MaxCapacity: 1,
      vpc,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      },
    })

    provider = new Provider(stack, "Provider", {
      vpc: vpc,
      cluster: cluster,
      secret: cluster.secret!,
    })
  })

  /**
   * Helper function to validate that all expected properties are present and correct
   */
  function validateCustomResourceProperties(
    template: Template,
    resourceType: string,
    expectedProperties: Record<string, any>
  ) {
    const resources = template.findResources("AWS::CloudFormation::CustomResource", {
      Properties: {
        Resource: resourceType,
      },
    })

    expect(Object.keys(resources)).toHaveLength(1)
    const resourceKey = Object.keys(resources)[0]!
    const resource = resources[resourceKey]!

    // Check each expected property individually
    for (const [key, expectedValue] of Object.entries(expectedProperties)) {
      if (typeof expectedValue === "object" && expectedValue?.name === "anyValue") {
        // Just check that the property exists
        expect(resource.Properties[key]).toBeDefined()
      } else if (typeof expectedValue === "object" && expectedValue?.name === "absent") {
        // Check that the property is absent
        expect(resource.Properties[key]).toBeUndefined()
      } else {
        // Check exact value
        expect(resource.Properties[key]).toEqual(expectedValue)
      }
    }

    // Verify service token is present and correct
    expect(resource.Properties.ServiceToken).toBeDefined()

    return resource
  }

  /**
   * Helper function to get all custom resources and verify their properties
   */
  function getAllCustomResourceProperties(template: Template) {
    const customResources = template.findResources("AWS::CloudFormation::CustomResource")
    const result: Record<string, any> = {}

    for (const [key, resource] of Object.entries(customResources)) {
      const resourceType = resource.Properties.Resource
      if (!result[resourceType]) {
        result[resourceType] = []
      }
      result[resourceType].push({
        key,
        properties: resource.Properties,
      })
    }

    return result
  }

  describe("Role resource properties", () => {
    test("role with database name includes all required properties", () => {
      new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        databaseName: "test_db",
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "role", {
        Resource: "role",
        ResourceId: "test_role",
        SecretArn: Match.anyValue(), // Secret ARN from provider
        PasswordArn: Match.anyValue(), // Generated secret ARN
        DatabaseName: "test_db",
        EnableIamAuth: false, // Converted from undefined to false by implementation
      })
    })

    test("role with IAM auth enabled", () => {
      new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        databaseName: "test_db",
        enableIamAuth: true,
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "role", {
        Resource: "role",
        ResourceId: "test_role",
        SecretArn: Match.anyValue(),
        PasswordArn: Match.anyValue(),
        DatabaseName: "test_db",
        EnableIamAuth: true,
      })
    })

    test("role without database name", () => {
      const database = new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
      })

      new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        database: database,
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "role", {
        Resource: "role",
        ResourceId: "test_role",
        SecretArn: Match.anyValue(),
        PasswordArn: Match.anyValue(),
        DatabaseName: "test_db", // Should come from database construct
        EnableIamAuth: false,
      })
    })

    test("role properties remain consistent across builds", () => {
      new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        databaseName: "test_db",
      })

      const template = Template.fromStack(stack)
      const resources = getAllCustomResourceProperties(template)

      // Snapshot test to catch unintended property changes
      expect(resources.role[0].properties).toMatchSnapshot()
    })
  })

  describe("Database resource properties", () => {
    test("database with owner includes all required properties", () => {
      const role = new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        databaseName: "test_db",
      })

      new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
        owner: role,
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "database", {
        Resource: "database",
        ResourceId: "test_db",
        SecretArn: Match.anyValue(),
        Owner: "test_role",
      })
    })

    test("database without owner", () => {
      new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "database", {
        Resource: "database",
        ResourceId: "test_db",
        SecretArn: Match.anyValue(),
        // Owner property is absent when not specified
      })
    })

    test("database properties remain consistent across builds", () => {
      new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
      })

      const template = Template.fromStack(stack)
      const resources = getAllCustomResourceProperties(template)

      expect(resources.database[0].properties).toMatchSnapshot()
    })
  })

  describe("Schema resource properties", () => {
    test("schema with role and database includes all required properties", () => {
      const database = new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
      })

      const role = new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        database: database,
      })

      new Schema(stack, "TestSchema", {
        provider: provider,
        schemaName: "test_schema",
        database: database,
        role: role,
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "schema", {
        Resource: "schema",
        ResourceId: "test_schema",
        SecretArn: Match.anyValue(),
        DatabaseName: "test_db",
        RoleName: "test_role",
      })
    })

    test("schema without role or database", () => {
      new Schema(stack, "TestSchema", {
        provider: provider,
        schemaName: "test_schema",
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "schema", {
        Resource: "schema",
        ResourceId: "test_schema",
        SecretArn: Match.anyValue(),
        // DatabaseName and RoleName are absent when not specified
      })
    })

    test("schema properties remain consistent across builds", () => {
      new Schema(stack, "TestSchema", {
        provider: provider,
        schemaName: "test_schema",
      })

      const template = Template.fromStack(stack)
      const resources = getAllCustomResourceProperties(template)

      expect(resources.schema[0].properties).toMatchSnapshot()
    })
  })

  describe("SQL resource properties", () => {
    test("sql with database and rollback includes all required properties", () => {
      const database = new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
      })

      new Sql(stack, "TestSql", {
        provider: provider,
        database: database,
        statement: "CREATE TABLE test (id INTEGER);",
        rollback: "DROP TABLE test;",
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "sql", {
        Resource: "sql",
        SecretArn: Match.anyValue(),
        DatabaseName: "test_db",
        Statement: "CREATE TABLE test (id INTEGER);",
        Rollback: "DROP TABLE test;",
      })
    })

    test("sql without database or rollback", () => {
      new Sql(stack, "TestSql", {
        provider: provider,
        statement: "CREATE TABLE test (id INTEGER);",
      })

      const template = Template.fromStack(stack)

      validateCustomResourceProperties(template, "sql", {
        Resource: "sql",
        SecretArn: Match.anyValue(),
        Statement: "CREATE TABLE test (id INTEGER);",
        // DatabaseName and Rollback are absent when not specified
      })
    })

    test("sql properties remain consistent across builds", () => {
      new Sql(stack, "TestSql", {
        provider: provider,
        statement: "CREATE TABLE test (id INTEGER);",
      })

      const template = Template.fromStack(stack)
      const resources = getAllCustomResourceProperties(template)

      expect(resources.sql[0].properties).toMatchSnapshot()
    })
  })

  describe("Imported Provider scenarios", () => {
    test("original provider Lambda has SECRET_ARN environment variable", () => {
      const template = Template.fromStack(stack)

      // Verify that the original provider's Lambda has SECRET_ARN in environment
      template.hasResourceProperties("AWS::Lambda::Function", {
        Environment: {
          Variables: {
            SECRET_ARN: Match.anyValue(),
          },
        },
      })
    })

    test("imported provider can be created with function ARN", () => {
      const importedStack = new cdk.Stack(app, "ImportedStack", {
        env: {
          account: "123456789012",
          region: "us-east-1",
        },
      })

      // This should not throw an error
      const importedProvider = Provider.fromProviderAttributes(
        importedStack,
        "ImportedProvider",
        {
          functionArn: "arn:aws:lambda:us-east-1:123456789012:function:test-provider",
          engine: DatabaseEngine.POSTGRES,
        }
      )

      // Verify the provider has the correct properties
      expect(importedProvider.engine).toBe(DatabaseEngine.POSTGRES)
      expect(importedProvider.serviceToken).toBeDefined()
    })

    test("imported provider can be created with function name", () => {
      const importedStack = new cdk.Stack(app, "ImportedStack2", {
        env: {
          account: "123456789012",
          region: "us-east-1",
        },
      })

      // This should not throw an error
      const namedProvider = Provider.fromProviderAttributes(
        importedStack,
        "NamedProvider",
        {
          functionName: "test-provider",
          engine: DatabaseEngine.MYSQL,
        }
      )

      // Verify the provider has the correct properties
      expect(namedProvider.engine).toBe(DatabaseEngine.MYSQL)
      expect(namedProvider.serviceToken).toBeDefined()
    })
  })

  describe("DSQL Provider scenarios", () => {
    let dsqlStack: cdk.Stack
    let dsqlCluster: dsql.CfnCluster
    let dsqlProvider: Provider

    beforeEach(() => {
      dsqlStack = new cdk.Stack(app, "DsqlStack", {
        env: {
          account: "123456789012",
          region: "us-east-1",
        },
      })

      const dsqlVpc = new ec2.Vpc(dsqlStack, "DsqlVpc", {
        subnetConfiguration: [
          {
            cidrMask: 28,
            name: "dsql",
            subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
          },
        ],
      })

      dsqlCluster = new dsql.CfnCluster(dsqlStack, "DsqlCluster")

      dsqlProvider = new Provider(dsqlStack, "DsqlProvider", {
        vpc: dsqlVpc,
        cluster: dsqlCluster,
      })
    })

    test("DSQL role creation does not include SecretArn", () => {
      new Role(dsqlStack, "DsqlRole", {
        provider: dsqlProvider,
        roleName: "dsql_role",
      })

      const template = Template.fromStack(dsqlStack)

      // DSQL uses IAM auth, so SecretArn should be absent
      validateCustomResourceProperties(template, "role", {
        Resource: "role",
        ResourceId: "dsql_role",
        PasswordArn: "", // DSQL doesn't use password
        EnableIamAuth: true, // DSQL always uses IAM auth
        // SecretArn should be absent for DSQL
        // DatabaseName should be absent for DSQL role without database
      })
    })
  })

  describe("Property validation across all resource types", () => {
    test("comprehensive property validation", () => {
      // Create all resource types
      const database = new Database(stack, "TestDatabase", {
        provider: provider,
        databaseName: "test_db",
      })

      const role = new Role(stack, "TestRole", {
        provider: provider,
        roleName: "test_role",
        database: database,
      })

      new Schema(stack, "TestSchema", {
        provider: provider,
        schemaName: "test_schema",
        database: database,
        role: role,
      })

      new Sql(stack, "TestSql", {
        provider: provider,
        database: database,
        statement: "CREATE TABLE test (id INTEGER);",
        rollback: "DROP TABLE test;",
      })

      const template = Template.fromStack(stack)
      const allResources = getAllCustomResourceProperties(template)

      // Verify all expected resource types are present
      expect(allResources).toHaveProperty("role")
      expect(allResources).toHaveProperty("database")
      expect(allResources).toHaveProperty("schema")
      expect(allResources).toHaveProperty("sql")

      // Verify each resource type has SecretArn (except for specific cases)
      for (const resourceType of ["role", "database", "schema", "sql"]) {
        const resources = allResources[resourceType]
        expect(resources).toHaveLength(1)
        expect(resources[0].properties.SecretArn).toBeDefined()
      }

      // Create comprehensive snapshot
      expect(allResources).toMatchSnapshot()
    })
  })
})
