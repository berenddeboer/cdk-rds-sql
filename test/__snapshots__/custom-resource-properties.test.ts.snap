// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Custom Resource Properties Database resource properties database properties remain consistent across builds 1`] = `
{
  "Resource": "database",
  "ResourceId": "test_db",
  "SecretArn": {
    "Ref": "ClusterSecretAttachment769E6258",
  },
  "ServiceToken": {
    "Fn::GetAtt": [
      "ProviderRdsSqlframeworkonEventD3D137B7",
      "Arn",
    ],
  },
}
`;

exports[`Custom Resource Properties Property validation across all resource types comprehensive property validation 1`] = `
{
  "database": [
    {
      "key": "TestDatabase",
      "properties": {
        "Resource": "database",
        "ResourceId": "test_db",
        "SecretArn": {
          "Ref": "ClusterSecretAttachment769E6258",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "ProviderRdsSqlframeworkonEventD3D137B7",
            "Arn",
          ],
        },
      },
    },
  ],
  "role": [
    {
      "key": "TestRolePostgresRole3AD07047",
      "properties": {
        "DatabaseName": "test_db",
        "EnableIamAuth": false,
        "PasswordArn": {
          "Ref": "TestRoleSecret0CBCDF24",
        },
        "Resource": "role",
        "ResourceId": "test_role",
        "SecretArn": {
          "Ref": "ClusterSecretAttachment769E6258",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "ProviderRdsSqlframeworkonEventD3D137B7",
            "Arn",
          ],
        },
      },
    },
  ],
  "schema": [
    {
      "key": "TestSchema",
      "properties": {
        "DatabaseName": "test_db",
        "Resource": "schema",
        "ResourceId": "test_schema",
        "RoleName": "test_role",
        "SecretArn": {
          "Ref": "ClusterSecretAttachment769E6258",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "ProviderRdsSqlframeworkonEventD3D137B7",
            "Arn",
          ],
        },
      },
    },
  ],
  "sql": [
    {
      "key": "TestSql",
      "properties": {
        "DatabaseName": "test_db",
        "Resource": "sql",
        "Rollback": "DROP TABLE test;",
        "SecretArn": {
          "Ref": "ClusterSecretAttachment769E6258",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "ProviderRdsSqlframeworkonEventD3D137B7",
            "Arn",
          ],
        },
        "Statement": "CREATE TABLE test (id INTEGER);",
      },
    },
  ],
}
`;

exports[`Custom Resource Properties Role resource properties role properties remain consistent across builds 1`] = `
{
  "DatabaseName": "test_db",
  "EnableIamAuth": false,
  "PasswordArn": {
    "Ref": "TestRoleSecret0CBCDF24",
  },
  "Resource": "role",
  "ResourceId": "test_role",
  "SecretArn": {
    "Ref": "ClusterSecretAttachment769E6258",
  },
  "ServiceToken": {
    "Fn::GetAtt": [
      "ProviderRdsSqlframeworkonEventD3D137B7",
      "Arn",
    ],
  },
}
`;

exports[`Custom Resource Properties SQL resource properties sql properties remain consistent across builds 1`] = `
{
  "Resource": "sql",
  "SecretArn": {
    "Ref": "ClusterSecretAttachment769E6258",
  },
  "ServiceToken": {
    "Fn::GetAtt": [
      "ProviderRdsSqlframeworkonEventD3D137B7",
      "Arn",
    ],
  },
  "Statement": "CREATE TABLE test (id INTEGER);",
}
`;

exports[`Custom Resource Properties Schema resource properties schema properties remain consistent across builds 1`] = `
{
  "Resource": "schema",
  "ResourceId": "test_schema",
  "SecretArn": {
    "Ref": "ClusterSecretAttachment769E6258",
  },
  "ServiceToken": {
    "Fn::GetAtt": [
      "ProviderRdsSqlframeworkonEventD3D137B7",
      "Arn",
    ],
  },
}
`;
